File: python/schemas/saju.py
"""
사주 분석 API 스키마 정의
Pydantic v2 모델
"""
from datetime import datetime
from enum import Enum
from pydantic import BaseModel, Field, field_validator


class Gender(str, Enum):
    """성별 (대운 방향 결정)"""
    MALE = "male"
    FEMALE = "female"


class CalculateRequest(BaseModel):
    """만세력 계산 요청"""
    birthDatetime: datetime = Field(
        ...,
        description="생년월일시 (ISO 8601 형식)",
        examples=["1990-05-15T14:30:00"]
    )
    timezone: str = Field(
        default="GMT+9",
        description="시간대 (예: GMT+9, GMT-5)",
        examples=["GMT+9"]
    )
    isLunar: bool = Field(
        default=False,
        description="음력 여부"
    )
    gender: Gender = Field(
        ...,
        description="성별 (대운 방향 결정)"
    )

    @field_validator('birthDatetime')
    @classmethod
    def validate_datetime(cls, v: datetime) -> datetime:
        """생년월일 범위 검증"""
        if v.year < 1900 or v.year > 2100:
            raise ValueError("지원 연도 범위: 1900-2100년")
        return v

    @field_validator('timezone')
    @classmethod
    def validate_timezone(cls, v: str) -> str:
        """시간대 형식 검증"""
        if not v.startswith("GMT"):
            raise ValueError("시간대 형식: GMT+9, GMT-5 등")
        return v

    model_config = {
        "json_schema_extra": {
            "example": {
                "birthDatetime": "1990-05-15T14:30:00",
                "timezone": "GMT+9",
                "isLunar": False,
                "gender": "male"
            }
        }
    }


class Pillar(BaseModel):
    """사주 기둥 (한 개)"""
    stem: str = Field(..., description="천간 (한자)", examples=["甲"])
    branch: str = Field(..., description="지지 (한자)", examples=["子"])
    element: str = Field(..., description="오행 (한자)", examples=["木"])


class Pillars(BaseModel):
    """사주 팔자 (4개 기둥)"""
    year: Pillar = Field(..., description="연주 (年柱)")
    month: Pillar = Field(..., description="월주 (月柱)")
    day: Pillar = Field(..., description="일주 (日柱)")
    hour: Pillar = Field(..., description="시주 (時柱)")


class DaewunItem(BaseModel):
    """대운 항목"""
    age: int = Field(..., description="시작 나이", examples=[1])
    stem: str = Field(..., description="천간", examples=["壬"])
    branch: str = Field(..., description="지지", examples=["午"])
    startYear: int = Field(..., description="시작 연도", examples=[1991])


class Jijanggan(BaseModel):
    """지장간 (각 지지에 숨어있는 천간)"""
    year: list[str] = Field(..., description="연주 지장간", examples=[["己", "丁"]])
    month: list[str] = Field(..., description="월주 지장간", examples=[["戊", "庚", "丙"]])
    day: list[str] = Field(..., description="일주 지장간", examples=[["癸"]])
    hour: list[str] = Field(..., description="시주 지장간", examples=[["己", "丁", "乙"]])


class CalculateResponse(BaseModel):
    """만세력 계산 응답"""
    pillars: Pillars = Field(..., description="사주 팔자")
    daewun: list[DaewunItem] = Field(..., description="대운 목록")
    jijanggan: Jijanggan = Field(..., description="지장간")

    model_config = {
        "json_schema_extra": {
            "example": {
                "pillars": {
                    "year": {"stem": "庚", "branch": "午", "element": "金"},
                    "month": {"stem": "辛", "branch": "巳", "element": "金"},
                    "day": {"stem": "甲", "branch": "子", "element": "木"},
                    "hour": {"stem": "辛", "branch": "未", "element": "金"}
                },
                "daewun": [
                    {"age": 1, "stem": "壬", "branch": "午", "startYear": 1991},
                    {"age": 11, "stem": "癸", "branch": "未", "startYear": 2001}
                ],
                "jijanggan": {
                    "year": ["己", "丁"],
                    "month": ["戊", "庚", "丙"],
                    "day": ["癸"],
                    "hour": ["己", "丁", "乙"]
                }
            }
        }
    }

File: src/types/saju.ts
/**
 * 사주(四柱) 관련 타입 정의
 */

/** 오행(五行) */
export type Element = 'wood' | 'fire' | 'earth' | 'metal' | 'water';

/** 오행 한자 매핑 */
export const ElementHanja: Record<Element, string> = {
  wood: '木',
  fire: '火',
  earth: '土',
  metal: '金',
  water: '水',
};

/** 천간(天干) */
export type HeavenlyStem =
  | 'gap' // 甲
  | 'eul' // 乙
  | 'byeong' // 丙
  | 'jeong' // 丁
  | 'mu' // 戊
  | 'gi' // 己
  | 'gyeong' // 庚
  | 'sin' // 辛
  | 'im' // 壬
  | 'gye'; // 癸

/** 천간 한자 매핑 */
export const HeavenlyStemHanja: Record<HeavenlyStem, string> = {
  gap: '甲',
  eul: '乙',
  byeong: '丙',
  jeong: '丁',
  mu: '戊',
  gi: '己',
  gyeong: '庚',
  sin: '辛',
  im: '壬',
  gye: '癸',
};

/** 지지(地支) */
export type EarthlyBranch =
  | 'ja' // 子
  | 'chuk' // 丑
  | 'in' // 寅
  | 'myo' // 卯
  | 'jin' // 辰
  | 'sa' // 巳
  | 'o' // 午
  | 'mi' // 未
  | 'sin' // 申
  | 'yu' // 酉
  | 'sul' // 戌
  | 'hae'; // 亥

/** 지지 한자 매핑 */
export const EarthlyBranchHanja: Record<EarthlyBranch, string> = {
  ja: '子',
  chuk: '丑',
  in: '寅',
  myo: '卯',
  jin: '辰',
  sa: '巳',
  o: '午',
  mi: '未',
  sin: '申',
  yu: '酉',
  sul: '戌',
  hae: '亥',
};

/** 사주 기둥 */
export interface Pillar {
  heavenlyStem: HeavenlyStem;
  earthlyBranch: EarthlyBranch;
  element: Element;
}

/** 사주 명반 (4개 기둥) */
export interface SajuPillars {
  year: Pillar; // 연주
  month: Pillar; // 월주
  day: Pillar; // 일주
  hour: Pillar; // 시주
}

/** 대운 정보 */
export interface Daewun {
  startAge: number;
  endAge: number;
  heavenlyStem: HeavenlyStem;
  earthlyBranch: EarthlyBranch;
  description?: string;
}

/** 사주 입력 데이터 */
export interface SajuInput {
  birthDate: Date;
  birthTime: string; // HH:mm 형식
  timezone: string; // GMT offset (예: 'GMT+9')
  isLunar: boolean;
  gender: 'male' | 'female';
}

/** 분석 집중 영역 */
export type FocusArea = 'wealth' | 'love' | 'career' | 'health' | 'overall';

/** 분석 집중 영역 한글 매핑 */
export const FocusAreaLabel: Record<FocusArea, string> = {
  wealth: '재물운',
  love: '연애운',
  career: '직장운',
  health: '건강운',
  overall: '종합운',
};

// ============================================
// Python API 응답 타입 (Task 8 추가)
// ============================================

/** 로딩 단계 */
export type LoadingStep =
  | 'manseryeok'
  | 'jijanggan'
  | 'ai_analysis'
  | 'visualization'
  | 'report'
  | 'complete';

/** 로딩 단계 한글 라벨 */
export const LoadingStepLabel: Record<LoadingStep, string> = {
  manseryeok: '만세력 계산',
  jijanggan: '지장간 추출',
  ai_analysis: 'AI 분석 진행',
  visualization: '시각화 생성',
  report: '최종 보고서 작성',
  complete: '완료',
};

/** Python API 응답용 Pillar (한자) */
export interface PillarHanja {
  stem: string; // 천간 한자 (甲, 乙, 丙...)
  branch: string; // 지지 한자 (子, 丑, 寅...)
  element: string; // 오행 한자 (木, 火, 土, 金, 水)
}

/** Python API 응답용 Pillars (4개 기둥) */
export interface PillarsHanja {
  year: PillarHanja;
  month: PillarHanja;
  day: PillarHanja;
  hour: PillarHanja;
}

/** Python API 대운 응답 */
export interface DaewunItem {
  age: number;
  stem: string;
  branch: string;
  startYear: number;
}

/** Python API 지장간 응답 */
export interface Jijanggan {
  year: string[];
  month: string[];
  day: string[];
  hour: string[];
}
